### ISSUE TITLE
Create/drop policies

### ISSUE THREAD

**Issue by jaulz** [tpetry/laravel-postgresql-enhanced#46]
What do you think about adding support for policies and row level security?

**Comment by jaulz** [comment]
I just came up with a macro in case someone else needs it:
````php

    Blueprint::macro('policy', function (
      string $policy,
      array $options = [],
    ) {
      /** @var \Illuminate\Database\Schema\Blueprint $this */
      $prefix = $this->prefix;
      $table = $this->table;

      $command = $this->addCommand(
        'policy',
        compact('prefix', 'table', 'policy', 'options')
      );
    });

    PostgresGrammar::macro('compilePolicy', function (
      Blueprint $blueprint,
      Fluent $command
    ) {
      /** @var \Illuminate\Database\Schema\Grammars\PostgresGrammar $this */
      $prefix = $command->prefix;
      $table = $command->table;
      $policy = $command->policy;
      $options = $command->options;

      $modifiers = [];
      foreach ($options as $key => $value) {
          $modifiers[] = match (true) {
              'as' === $key => "as {$value}",
              'for' === $key => "for {$value}",
              'to' === $key => "to {$value}",
              'using' === $key => "using ({$value})",
              'check' === $key => "with check ({$value})",
              default => throw_if(true, message: "Unknown option '{$key}'."),
          };
      }
      $sqlModifiers = implode(' ', $modifiers);

      return [
        sprintf('CREATE POLICY %s ON %s%s %s', $policy, $prefix, $table, $sqlModifiers),
      ];
    });
````

And then similar for drop:
````php
    Blueprint::macro('dropPolicy', function (
      string $policy,
    ) {
      /** @var \Illuminate\Database\Schema\Blueprint $this */
      $prefix = $this->prefix;
      $table = $this->table;

      $command = $this->addCommand(
        'dropPolicy',
        compact('prefix', 'table', 'policy')
      );
    });

    PostgresGrammar::macro('compileDropPolicy', function (
      Blueprint $blueprint,
      Fluent $command
    ) {
      /** @var \Illuminate\Database\Schema\Grammars\PostgresGrammar $this */
      $prefix = $command->prefix;
      $table = $command->table;
      $policy = $command->policy;

      return [
        sprintf('DROP POLICY IF EXISTS %s ON %s%s', $policy, $prefix, $table),
      ];
    });
````

Maybe we can incorporate it in this library as well 😊

**Comment by tpetry** [comment]
I am planning to support RLS, but I don't know yet what the API will look like. I am unsure whether I want to make it look like the SQL syntax or some higher-level API. Also working with RLS within Laravel is not that easy so I am thinking about whether some helpers need to be added to make development easier - and make it a risk-free approach.

**Comment by Radiergummi** [comment]
@tpetry Are you still considering this? I've been thinking about RLS in Laravel for a while now, but the most annoying part by far is making it easy to set the current security context in front of all queries, further complicated by things like poolers or Octane.  
However, the way it could work, I think, would be by letting the driver make sure every query is wrapped in a transaction, and set the security context (via `set_config(..., true)`) when it begins.  
That way, users don't have to wrap entire the request handling in one huge, implicit transaction; queries are still fully isolated, but the security context is set for all of them; and users don't have to modify their existing queries to accommodate.

It sure adds some overhead, but for the security guarantees it gives you I think that's fine - this feature would obviously be opt-in anyway.

Would you be interested in adding such a feature to the library in the first place? If so, I'd like to offer creating a prototype to see if/how that could work.

**Comment by tpetry** [comment]
RLS is still on my wishlist. But its a complicated feature with a lot of implications. So when I support it, it will only be the migration part. How to use it should be an extra package as there are many ways to use it and I don't want to enforce a special way.

For example, how to ensure that requests are applying the RLS tenant_id. This has an impact on writing the controllers with transaction encapsulation. Probably, this should be a middleware so its easy to use. Then a job is dispatched and you must know the tenant_id there too - or not if jobs are allowed to run with more privileges. The Laravel context could be helpful.

Something like the login needs access to all users. So either a database connection with full rights must be used, one with only access to the users table or best a custom function which can only do lookups and not access all data. Same for all other auth.

Its complicated. And there many different ways to do it. I would have to enforce structuring an app in a very specific way. But this database driver shouldn't enforce it. Its agnostic on how you want to use any features.

**Comment by Radiergummi** [comment]
I agree that this should be a standalone package, although there's a fair bit of entanglement between the schema and the application facilities. That said, I built a proof of concept package over the weekend. If you're interested, I'd be very glad to get your opinion: https://github.com/Radiergummi/laravel-rls

It should work with any tenancy setup, allows to use a single or separate database connections, and both wrapping the entire request in a transaction or all individual queries. Quite happy with it, but not used in a real application yet.

**Comment by tpetry** [comment]
I've looked through it. And from my experience working with RLS, its a good starting point. But I wouldn't say it can/should be used if not an entire project has been built on top of it. You will experience a lot of shortcomings on bigger projects. Just from a quick glance I see these issues:
* No granular configuration options for the RLS policy. At first, they are not needed but over time you will need them. Especially when e.g. user roles are also mapped into RLS that a user can only read but not change rows etc.
* Not sure whether the `Rls::defineContext` works in reality on bigger projects. Sometimes a session/transaction even has multiple RLS properties because a simple tenant_id is not enough.
* Is the RLS context of a http request passed to a queued job/event? That is essential - and sometimes its event not wanted. So again, a much more complex implementation is needed that e.g. everything queueable can define whether it wants to retain the RLS context.

But things I didnt think about and are really nice:
* Passing a reason why a query is run without RLS is great for documentation reasons. But not sure whether these should be logged, e.g. auth functionality always must run without RLS or very special constructs.
* Providing testing functionality is great. Its so easy to break some RLS stuff in migrations.
